provider/segmenter/src/lib.rs

```diff
@@ -0,0 +1,54 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! `icu_provider_segmenter` contains implementations of the [`ICU4X`] [data provider] interface
+//! based on Unicode properties and TOML files implementing [Unicode Standard Annex #14][UAX14] and
+//! [Unicode Standard Annex #29][UAX29] breaking rules.
+//!
+//! **Important:** This provider implementation is not optimized for production use. It is much more
+//! efficient if you use [`FsDataProvider`] or [`StaticDataProvider`] instead.
+//!
+//! # Examples
+//!
+//! ```
+//! use icu_provider_segmenter::SegmenterRuleProvider;
+//! let provider = SegmenterRuleProvider::try_new("/path/to/data/directory")
+//!     .expect_err("Specify a real directory in the line above");
+//! ```
+//!
+//! # Exporting data
+//!
+//! To generate the data required by the segmenters, run `cargo make testdata` from the top level.
+//!
+//! [`ICU4X`]: ../icu/index.html
+//! [data provider]: icu_provider
+//! [UAX14]: https://www.unicode.org/reports/tr14/
+//! [UAX29]: https://www.unicode.org/reports/tr29/
+//! [`FsDataProvider`]: ../icu_provider_fs/struct.FsDataProvider.html
+//! [`StaticDataProvider`]: ../icu_provider_blob/struct.StaticDataProvider.html
+
+use icu_provider::fork::by_key::MultiForkByKeyProvider;
+use icu_provider::iter::IterableDynProvider;
+use icu_provider::prelude::*;
+use std::path::{Path, PathBuf};
+
+mod transform;
+
+pub use crate::transform::SegmenterRuleProvider;
+
+/// Returns the absolute path to the directory containing rule break data.
+pub fn break_data_root() -> PathBuf {
+    PathBuf::from(std::env!("CARGO_MANIFEST_DIR")).join("data")
```

Comment:
When adding LSTM (JSON) and dictionary (currently this is binary data from ICU4C), do we store it into same directory?

If you think that others are stored into `provider/testdata/data/json/segmenter` via JSON format directory, please ignore my comment.
I imagine that the following.
- `data/json/segmenter/lstm/th.json` or `data/json/segmenter_lstm/th.json`
- `data/json/segmenter/char16trie/th.json` or `data/json/segmenter_char16trie/th.json` (convert from binary to JSON)
- etc

Reply:
Hmm, I take a closer look at how Unicode properties testdata are structured.

For ICU4C generated raw toml data, they are placed in `provider/testdata/data/uprops`, and icu4x-datagen generates testdata under `provider/testdata/data/json/props` and a binary format under `provider/testdata/data/testdata.postcard`.

Also, note that the [Resource key](https://unicode-org.github.io/icu4x-docs/doc/icu_provider/prelude/struct.ResourceKey.html) has a restrict format, and it determines the generated data path and file name. 

If LSTM and dictionary data's raw format and runtime format is similar or even identical, the provider to transform the raw data to the runtime data can be trivial, but I guess we still need to write such a provider for `icu4x-datagen`?

I image we can put the segmenter raw data (rule break TOML, raw LSTM json, and ICU4C generated dictionary data) under `provider/testdata/data/segmenter`, and the the resource key -> generate data path will be like:
 - `LineBreakDataV1Marker = "segmenter/line@1"` -> `provider/testdata/data/json/segmenter/line@1.json`
 - `ThaiDictionaryBreakDataV1Marker = segmenter/thai_dict@1` -> `provider/testdata/data/json/segmenter/thai_dict@1.json`
 - `ThaiLstmBreakDataV1Marker = segmenter/thai_lstm@1` -> `provider/testdata/data/json/segmenter/thai_lstm@1.json`

(Yes, I think we need different key per language for dictionary and lstm)

I know there is a plan to make icu4x-datagen more friendly to the user, so the data path might be subjected to change. I'll wait @sffc's review and follow his opinion on how we should structure the segmenter data so that they integrate better.

Reply:
What @aethanyc wrote above looks correct to me.

Reply:
OK, I make sense.